### PR TITLE
tmc5160.py does not update CS values

### PR DESCRIPTION
**Pull request by HonestBrothers**
This uses the tuning spreadsheet to calculate CS values versus globalscaler.

pg. 74 of the datasheet shows that Irms is calculated by the tmc5160:

Irms = (globalscaler/256) * ((CS+1)/32) * (Vfs/Rsense) * (1/sqrt(2))

This leaves us with one equation and two unknowns, which results in an unsolvable equation. The current code in Klipper then ignores the CS value and solves the above equation for global scaler:

globalscaler = int((current * 256. * math.sqrt(2.) * self.sense_resistor / VREF) + .5)

Effective what this does, since multiplying by 1 results in the same number that we started with, is sets ((CS+1)/32) = 1. This means that CS is 31. 

Irms = (globalscaler/256) * ((31+1)/32) * (Vfs/Rsense) * (1/sqrt(2))

Irms = (globalscaler/256) * 1 * (Vfs/Rsense) * (1/sqrt(2))

Klipper then back calculates the CS value. The problem with this is the way the driver implements this is it uses BOTH CS and Globalscaler in the same equation. So we can't back calculate because those maths don't math. One equation, two unknown functions are unsolvable, we would need two equations for this.

### CODE REVIEW

**Comment by richfelker** [comment]
Can you clarify what's functionally wrong right now? Is the actual current significantly different from the requested current because of this bug? Are there other symptoms?

**Comment by HonestBrothers** [comment]
The PR I raised on the main branch outlines the issue. Essentially the current is slightly different (not enough to make a difference), but the big issue is the chopper hysteresis can't be properly set in many circumstances without access to change CS. 

Conversation with Dmitry basically boiled down to the fact that we may just need to expose the CS value as a parameter we can set, instead of auto calculating it. Auto calculating will screw up every tmc5160 driver settings. 

But, the way it's set right now, the settings are likely already screwed up. 

https://github.com/Klipper3d/klipper/pull/6644

**Comment by bwnance** [comment]
@HonestBrothers i'm not opposed to adopting this change, but i don't want to break existing chopper settings. I see two options:

1. continue to auto-calc, but use a config flag in the tmc section to determine which algorithm to use
2. support setting of the CS parameter - in the absence of the CS parameter, however, auto-calc it using the current (old) method.

File: klippy/extras/tmc5160.py [hunk 1]
```diff
@@ -258,6 +262,7 @@ def __init__(self, config, mcu_tmc):
         self.sense_resistor = config.getfloat(
             "sense_resistor", 0.075, above=0.0
         )
+        self.cs = config.getint('tmc_cs', 31, 0)
```
Comment (bwnance) [inline klippy/extras/tmc5160.py:265]: use a keyword arg for the minval - `self.cs = config.getint('tmc_cs', 31, minval=0)`

File: klippy/extras/tmc5160.py [hunk 2]
```diff
@@ -266,30 +271,23 @@ def __init__(self, config, mcu_tmc):
...
-        cs = int(
-            (current * 256.0 * 32.0 * math.sqrt(2.0) * self.sense_resistor)
-            / (globalscaler * VREF)
-            - 1.0
-            + 0.5
-        )
-        return max(0, min(31, cs))
+    def _calc_current_bits(self, current):
+        cs = self.cs
+        return max(16, min(31, cs))
```
Comment (bwnance) [inline klippy/extras/tmc5160.py:285]: i think you can just do maxval=31 in the `config.getint` line in the init, and get rid of the min here.

Comment (bwnance) [reply klippy/extras/tmc5160.py:285]: also, what's `16` for? is the idea that 16 is the minimum val for cs? if so, we can just do minval=16 in the getint.

**Review by bwnance — CHANGES_REQUESTED** [review]

File: klippy/extras/tmc5160.py [hunk 3]
```diff
@@ -3,6 +3,10 @@
 # Copyright (C) 2018-2019  Kevin O'Connor <kevin@koconnor.net>
 #
 # This file may be distributed under the terms of the GNU GPLv3 license.
+#
+# Modified by Honest Brothers to correct CS calcs (c) 2024
```
Comment (bwnance) [inline klippy/extras/tmc5160.py:7]: also, we're avoiding doing headers for attribution - the commit will be in your name so you'll get attribution for the contributions that way (in the git blame)

File: klippy/extras/tmc5160.py (hunk 2 above)
Comment (HonestBrothers) [reply klippy/extras/tmc5160.py:285]: Yes. Changing.

Comment (HonestBrothers) [reply klippy/extras/tmc5160.py:285]: Some of the code below could change to be shrunk, I'll work on that. I'm shit at python/coding, so keep that in mind. I just break stuff until it's not broken anymore. 

For instance _calc_current_bits passes current for no reason, other than to pass it.

Comment (HonestBrothers) [reply klippy/extras/tmc5160.py:285]: I made this change, but we need a line somewhere that autosets to 31 if tmc_cs isn't defined.

**Comment by dalegaard** [comment]
Hi,

So there seems to be some misunderstandings about why this change is needed(especially in the mainline discussion). Basically it boils down to chopper hysteresis settings only being applied to the CS term only. In cases where extreme hysteresis is needed in relation to the CS value, reducing CS can get the hysteresis values back to within limits. The lost current can then be made up by setting globalscaler higher to match.

Only very few motors should require this, but I think this is overall a good change. Having CS=31 be the default, so CS never becomes 30 anymore, seems fine to me.

Best regards,
Lasse

**Comment by HonestBrothers** [comment]
> Hi,
> 
> So there seems to be some misunderstandings about why this change is needed(especially in the mainline discussion). Basically it boils down to chopper hysteresis settings only being applied to the CS term only. In cases where extreme hysteresis is needed in relation to the CS value, reducing CS can get the hysteresis values back to within limits. The lost current can then be made up by setting globalscaler higher to match.
> 
> Only very few motors should require this, but I think this is overall a good change. Having CS=31 be the default, so CS never becomes 30 anymore, seems fine to me.
> 
> Best regards, Lasse

Feel free to comment on the mainline branch, if you want. I end up going on too many tangents. 

I don't think the change will be needed often, but those that want to run high end motors will benefit. Defaulting to 31, in my opinion is fine, because under the current implementation you'd be waaaay outside the boundaries of the driver when CS switches to 30. Like scarecrow keeps saying, you really need to change RSense or switch drivers at that point. 

I don't believe the mainline code would behave in the way Trinamic intended it to, even if the code didn't switch to CS=30 at Globalscaler <= 32. That's one of the tangents I keep going on, but have probably poorly explained.

**Comment by MSzturc** [comment]
> I don't think the change will be needed often, but those that want to run high end motors will benefit. Defaulting to 31, in my opinion is fine, because under the current implementation you'd be waaaay outside the boundaries of the driver when CS switches to 30. Like scarecrow keeps saying, you really need to change RSense or switch drivers at that point.
> 
> I don't believe the mainline code would behave in the way Trinamic intended it to, even if the code didn't switch to CS=30 at Globalscaler <= 32. That's one of the tangents I keep going on, but have probably poorly explained.

That's 100% true for TMC2209 drivers, but in my point of view it does not fit for TMC5160

Most TMC5160 drivers come with low Rsens resistors, like 0.050 or 0.022 ohm and since people use them to drive more premium motors like 2504er, 2804er, CM06 or CM08, all these motors have a coil resistance/inductance that **requires** a cs lower than 31.

The current TMC5160 implementation makes it impossible to use TMC5160 with klipper properly. I did over the last few weeks multiple tests and the results where shocking. On my current printer configuration (4x2804 on a BTT Kraken) I was not able to reach speeds over 1200mm/s at 130k acceleration with the standard klipper behaviour (cs=31), Compared to a configuration with cs=9 that setup tapped out at 1600mm/s at 280k acceleration which is more than double.

Also I would prefer to implement an auto calculation of cs since honest every stepper on a tmc5160 will profit from that optimisation I did a lot of calculation and even the cheapest motors for $4-5 combined with the worst tmc5160 driver (1st gen with 0.075 rsens) will need lower cs values in every high performance and every silent szenario

**Comment by HonestBrothers** [comment]
@MSzturc that's some really good data. 

The problem with auto-calculating is it screws up all the existing implementations of the 5160 code deployed. Suddenly everyone will need to recalibrate their drivers.

**Comment by Andor** [comment]
@HonestBrothers is it possible to add option "auto" for this parameter in the config?
Or maybe "magic" value like `-1` which will mean `auto` if adding a string value instead of integer one is difficult.

**Comment by MSzturc** [comment]
> @MSzturc that's some really good data.
> 
> The problem with auto-calculating is it screws up all the existing implementations of the 5160 code deployed. Suddenly everyone will need to recalibrate their drivers.

I think this point is debatable. With the current implementation the people had to choose settings to compenbsate the flaws in the implementation. Those settings increase the power dissipation for the driver. To use a metaphor the drive a car with a gas pedal that only knows 0 and max and since that is an issue they choose to limit the fuel that gets injected to the engine (globalscaler). So basically in the current setup people roasting their drivers, shorten their lifespan, increase artefacts on their prints.

I personally would classify this bug as critical and therefore I would suggest implementing a non backward compatible implementation. But as I said it's debatable and I also understand the standpoint of the klipper guys maneuver around the shitstorm that change would cause. Therefore suggested an auto version that defaults to 31 in the codereview

**Comment by HonestBrothers** [comment]
@MSzturc I agree that the power dissipates in the driver, but immediate harm might come to the motors/drivers if we switch to auto-calculating all the sudden. 

I've blown one of these drivers by switching settings and making no hardware changes.

**Comment by MSzturc** [comment]
> @MSzturc I agree that the power dissipates in the driver, but immediate harm might come to the motors/drivers if we switch to auto-calculating all the sudden.
> 
> I've blown one of these drivers by switching settings and making no hardware changes.

I'm pretty sure that changing the CS would not damage your drivers since the global scaler limits the potential damage, but on the other hand I don't know where all the chinese board manufacturers are cutting corners on their low budget printer boards, so defaulting to 31 might be a good idea for klipper standard,

**Comment by CarVac** [comment]
Damage can occur from ringing and from too-short break-before-make time (which exists to prevent the driver from shorting the mosfets out).

**Comment by HonestBrothers** [comment]
I updated everything.

1. tmc_cs: 0
Sets the system to auto-calculate
2. tmc_cs: 1-30
Lets the user choose the tmc_cs value
3. tmc_cs: not defined or 31
Default condition. 

CS value is updated for ihold if defined. 

Code should be ready to merge.

**Comment by MSzturc** [comment]
> I updated everything.
> 
> 1. tmc_cs: 0
>    Sets the system to auto-calculate
> 2. tmc_cs: 1-30
>    Lets the user choose the tmc_cs value
> 3. tmc_cs: not defined or 31
>    Default condition.
> 
> CS value is updated for ihold if defined.
> 
> Code should be ready to merge.

It might be a bit pedantic but it would be easier for people to find that setting when it's part of the klipper naming convention, starting with **driver_**

Also even it's not a realistic szenario technically cs = 0 is a valid value regarding to the documentation of TMC

**Comment by HonestBrothers** [comment]
@MSzturc I will update to fix the naming convention. 

Not sure how I feel about CS=0. I suppose I should code in accordance to the datasheet. I will change this to -1.

**Comment by MSzturc** [comment]
> @MSzturc I will update to fix the naming convention.
> 
> Not sure how I feel about CS=0. I suppose I should code in accordance to the datasheet. I will change this to -1.

Yeah I know, it's unrealistic that cs=0 will be ever needed but im sure that in the future there will be a person looking into the code and being confused, so at least add a comment

**Comment by HonestBrothers** [comment]
Changed.

driver_cs: -1
  Sets the system to auto-calculate
driver_cs: 0-30
   Lets the user choose the tmc_cs value
driver_cs: not defined or 31
   Default condition.

File: klippy/extras/tmc5160.py (hunk 2 above)
Comment (HonestBrothers) [reply klippy/extras/tmc5160.py]: @bwnance I believe this works the way that it should now. Uses consistent naming conventions, allows auto calculations, defaults to 31.

**Comment by CarVac** [comment]
Reminder to remove the attribution per @bwnance 's requested change.

**Comment by HonestBrothers** [comment]
@CarVac didn't realize it was still there. I reverted to an old version before I uploaded.

**Review by MSzturc — APPROVED** [review]
Looks good to me!

File: klippy/extras/tmc5160.py [hunk 4]
```diff
@@ -257,21 +259,18 @@ def __init__(self, config, mcu_tmc):
...
-                This parameter will be mandatory in future versions.
-                Specify the parameter to resolve this warning""",
+                """,
                 self.name,
                 "sense_resistor",
             )
 
-        self.sense_resistor = config.getfloat(
-            "sense_resistor", 0.075, above=0.0
-        )
```
Comment (rogerlz) [inline klippy/extras/tmc5160.py:274]: why this change? I still want to keep it as a warning for now

File: klippy/extras/tmc5160.py [hunk 5]
```diff
@@ -280,30 +279,30 @@ def __init__(self, config, mcu_tmc):
...
+    def _calc_current_bits(self, current):
+        if self.cs == -1:
+            Ipeak = current * math.sqrt(2)
+            Rsens = self.sense_resistor
+            cs = int(math.ceil(Rsens * 32 * Ipeak / 0.32) - 1)
+        elif self.cs < 31:
+            cs = self.cs
+        else:
+            cs = 31
```
Comment (rogerlz) [inline klippy/extras/tmc5160.py:299]:
```
if self.cs == -1:
    Ipeak = current * math.sqrt(2)
    Rsens = self.sense_resistor
    cs = int(math.ceil(Rsens * 32 * Ipeak / 0.32) - 1)
else:
    cs = self.cs     
```

Comment (rogerlz) [reply klippy/extras/tmc5160.py:299]: `getint()` is handling the default and maxval

File: klippy/extras/tmc5160.py (hunk 4 above)
Comment (HonestBrothers) [reply klippy/extras/tmc5160.py:274]: I don't think we should auto set the Rsense. Dangerous. It's default mainline Klipper behavior, but bad practice.

Comment (rogerlz) [reply klippy/extras/tmc5160.py:274]: I agree, and that's why there is a warning about it. we don't want to break configs out there

Comment (CarVac) [reply klippy/extras/tmc5160.py]: That's the opposite of what you want.

A high *actual* sense resistor reduces the max current. Telling Klipper that the sense resistor is larger than the actual sense resistor will result in over-current.

Comment (HonestBrothers) [reply klippy/extras/tmc5160.py]: Should the physical resistor only scale VRef differently? V drop across the resistor will vary based on Rs, which will screw up the comparator. The Imot sine wave would be all sorts of screwed up, but Imot peak would be lower based on Rsense? At least that's what I gathered from pg. 74. 

I could be wrong.

Comment (CarVac) [reply klippy/extras/tmc5160.py]: Higher physical rsense than expected: you get the desired voltage drop at a lower current than desired. This is safe.

Lower physical rsense than expected: you get the desired voltage drop at a higher current than desired. This is unsafe, and given that many TMC5160 boards have much lower rsense than the assumed 75 mOhm, you end up with overcurrent by default.

Comment (HonestBrothers) [reply klippy/extras/tmc5160.py]: Thanks! I was definitely backwards.

Comment (CarVac) [reply klippy/extras/tmc5160.py]: Agreed. Preserving people's configurations is one thing, but if they aren't configuring rsense they ought to be informed that they are making a mistake.

If we do break the present default, I think it's worth having the error message specify that the _previous_ rsense default was 0.075 ohm.

**Comment by HonestBrothers** [comment]
@Zeanon 

> @HonestBrothers Do you know how to apply the same thing to tmc2240 Since they suffer from the same issue

I will look at the code in a couple weeks and give it a go. Should be fairly trivial. I don't own these drivers though, so I may need your support in testing. 

I'm in the middle of moving houses and my entire system is down for the near term. Got to get my old house fixed and on the market, and then set up my office and I'll have some time to work on it.

**Comment by rogerlz** [comment]
Hi, 

I am closing this one in favour of a new implementation here: https://github.com/DangerKlippers/danger-klipper/pull/374

**Pull request closed without merging**